go-controller/pkg/cluster/node.go

```diff
@@ -215,22 +215,101 @@ func (cluster *OvnClusterController) initGateway(
 	}
 
 	err = util.GatewayInit(clusterIPSubnet, nodeName, ipAddress, "",
-		cluster.GatewayIntf, cluster.GatewayNextHop, subnet)
+		cluster.GatewayBridge, cluster.GatewayNextHop, subnet)
+	if err != nil {
+		return err
+	}
+
+	if !cluster.NodePortEnable {
+		// Program cluster.GatewayIntf to let non-pod tarffic to go to host
+		// stack
+		err = cluster.addDefaultConntrackRules()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (cluster *OvnClusterController) addDefaultConntrackRules() error {
+	patchPort := "k8s-patch-" + cluster.GatewayBridge + "-br-int"
+	// Get ofport of pathPort
+	ofportPatch, stderr, err := util.RunOVSVsctl("--if-exists", "get",
+		"interface", patchPort, "ofport")
+	if err != nil {
+		return fmt.Errorf("Failed to get ofport of %s, stderr: %q, error: %v",
+			patchPort, stderr, err)
+	}
+
+	// Get ofport of physical interface
+	ofportPhys, stderr, err := util.RunOVSVsctl("--if-exists", "get",
+		"interface", cluster.GatewayIntf, "ofport")
+	if err != nil {
+		return fmt.Errorf("Failed to get ofport of %s, stderr: %q, error: %v",
+			cluster.GatewayIntf, stderr, err)
+	}
+
+	// table 0, packets coming from pods headed externally. Commit connections
+	// so that reverse direction goes back to the pods.
+	_, stderr, err = util.RunOVSOfctl("add-flow", cluster.GatewayBridge,
+		fmt.Sprintf("priority=100, in_port=%s, ip, "+
+			"actions=ct(commit, zone=%d), output:%s",
+			ofportPatch, config.ConntrackZone, ofportPhys))
+	if err != nil {
+		return fmt.Errorf("Failed to add openflow flow to %s, stderr: %q, "+
+			"error: %v", cluster.GatewayBridge, stderr, err)
+	}
 
-	return err
+	// table 0, packets coming from external. Send it through conntrack and
+	// resubmit to table 1 to know the state of the connection.
+	_, stderr, err = util.RunOVSOfctl("add-flow", cluster.GatewayBridge,
+		fmt.Sprintf("priority=50, in_port=%s, ip, "+
+			"actions=ct(zone=%d, table=1)", ofportPhys, config.ConntrackZone))
+	if err != nil {
+		return fmt.Errorf("Failed to add openflow flow to %s, stderr: %q, "+
+			"error: %v", cluster.GatewayBridge, stderr, err)
+	}
+
+	// table 1, established and related connections go to pod
+	_, stderr, err = util.RunOVSOfctl("add-flow", cluster.GatewayBridge,
+		fmt.Sprintf("priority=100, table=1, ct_state=+trk+est, "+
+			"actions=output:%s", ofportPatch))
+	if err != nil {
+		return fmt.Errorf("Failed to add openflow flow to %s, stderr: %q, "+
+			"error: %v", cluster.GatewayBridge, stderr, err)
+	}
+	_, stderr, err = util.RunOVSOfctl("add-flow", cluster.GatewayBridge,
+		fmt.Sprintf("priority=100, table=1, ct_state=+trk+rel, "+
+			"actions=output:%s", ofportPatch))
+	if err != nil {
+		return fmt.Errorf("Failed to add openflow flow to %s, stderr: %q, "+
+			"error: %v", cluster.GatewayBridge, stderr, err)
+	}
+
+	// table 1, all other connections go to the bridge interface.
+	_, stderr, err = util.RunOVSOfctl("add-flow", cluster.GatewayBridge,
+		"priority=0, table=1, actions=output:LOCAL")
```

Comment:
I presume this means actions=NORMAL

Reply:
actions=NORMAL means to treat it is a learning switch. actions=LOCAL means to send the packet to the interface in the bridge which has the same name as the bridge. NORMAL can create confusions as we have 2 interfaces with the same MAC address and likely more broadcasts with slower performance (one local bridge and one which is OVN router interface) - so this makes it explicit that we send it to LOCAL.  There is a lower priority rule with NORMAL which exists because we need SSH connection to work when we move from nic to bridge.